File: Easy/0160_intersection-of-two-linked-lists/intersection-of-two-linked-lists.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val int
	Next *ListNode
}

func getIntersectionNode(headA, headB *ListNode) *ListNode {
	if headA == nil || headB == nil {
		return nil
	}
	pa, pb := headA, headB
	for pa != pb {
		if pa != nil {
			pa = pa.Next
		} else {
			pa = headB
		}
		if pb != nil {
			pb = pb.Next
		} else {
			pb = headA
		}
	}
  return pa
}

func buildList(values []int) *ListNode {
	dummy := &ListNode{}
	curr := dummy
	for _, v := range values {
		curr.Next = &ListNode{Val: v}
		curr = curr.Next
	}
	return dummy.Next
}

func buildAttachLists(listA, listB []int, intersectA, intersectB int) (*ListNode, *ListNode) {
	headA := buildList(listA)
	headB := buildList(listB)

	if intersectA == - 1 || intersectB == -1 {
		return headA, headB
	}

	intersectNode := headA
	for i := 0; i < intersectA && intersectNode != nil; i++ {
		intersectNode = intersectNode.Next
	}

	attachPoint := headB
	for i := 0; i < intersectB && attachPoint != nil; i++ {
		attachPoint = attachPoint.Next
	}

	if attachPoint != nil {
		attachPoint.Next = intersectNode
	}
	return headA, headB
}


type testCase struct {
	lA []int
	lB []int
	interA int
interB int
}

func main() {
	tests := []testCase{
		{[]int{4, 1, 8, 4, 5}, []int{5, 6, 1, 8, 4, 5}, 2, 3},
		{[]int{1, 9, 1, 2, 4}, []int{3, 2, 4}, 3, 1},
		{[]int{2, 6, 4}, []int{1, 5}, 3, 2},
	}

	for _, t := range tests {
		fmt.Printf("---\nlistA: %v listB: %v ", t.lA, t.lB)
		fmt.Printf("intersectA: %d intersectB: %d\n", t.interA, t.interB)

		headA, headB := buildAttachLists(t.lA, t.lB, t.interA, t.interB)
		node := getIntersectionNode(headA, headB)

		if node != nil {
			fmt.Printf("intersectNode: %d\n", node.Val)
		} else {
			fmt.Printf("intersectNode: nil\n")
		}
	}
}

File: Easy/0160_intersection-of-two-linked-lists/intersection-of-two-linked-lists.py
#!/bin/python3

from typing import List, Optional, Tuple
from dataclasses import dataclass

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        if not headA or not headB:
            return None

        pa, pb = headA, headB

        while pa != pb:
            pa = pa.next if pa else headB
            pb = pb.next if pb else headA

        return pa

def buildAttachLists(listA: List[int], listB: List[int], intersectA: int, intersectB: int) -> Tuple[Optional[ListNode], Optional[ListNode]]:
    def buildList(values: List[int]) -> Optional[ListNode]:
        dummy = ListNode(0)
        current = dummy
        for val in values:
            current.next = ListNode(val)
            current = current.next
        return dummy.next

    headA = buildList(listA)
    headB = buildList(listB)

    if intersectA == -1 or intersectB == -1:
        return headA, headB

    intersect_node = headA
    for _ in range(intersectA):
        if intersect_node:
            intersect_node = intersect_node.next

    attach_point = headB
    for _ in range(intersectB):
        if attach_point:
            attach_point = attach_point.next

    if attach_point:
        attach_point.next = intersect_node

    return headA, headB

@dataclass
class testCase:
    listA: List[int]
    listB: List[int]
    intersectA: int
    intersectB: int

def main():
    tests: List[testCase] = [
        testCase([4,1,8,4,5], [5,6,1,8,4,5], 2, 3),
        testCase([1,9,1,2,4], [3,2,4], 3, 1),
        testCase([2,6,4], [1,5], 3, 2),
    ]

    for t in tests:
        print(f"---\nlistA: {t.listA} listB: {t.listB}", end=" ")
        print(f"intersectA: {t.intersectA} intersectB: {t.intersectB}")
        headA, headB = buildAttachLists(t.listA, t.listB, t.intersectA, t.intersectB)
        node = Solution().getIntersectionNode(headA, headB)
        print(f"intersectNode: {node.val if node else None}")

if __name__ == "__main__":
    main()
